docs/DafnyRef/Refinement.md

```diff
@@ -92,32 +92,133 @@ TODO
 
-To indicate it is a refining declaration, a refining `const` declaration
-contains an ellipsis in this syntax:
-```grammar
-"const" { Attribute } CIdentType "..." [ ":=" Expression ]
-```
+A refining module can also introduce new `const` declarations that do
+not exist in the refinement parent.
 
 ## 21.5. Method declarations
 
-TODO
+A child `method` definition may refine a parent `method` declaration or
+definition to
+
+* provide a body missing in the parent,
+* replace the body of the parent with a new, semantically-compatible
+  body,
+* strengthen the postcondition of the parent method by adding one or more
+  `ensures` clauses,
+* provide a more deterministic version of a non-deterministic parent
+  body, or
+* amend the body of the parent method with additional statements.
+
+The type signature of a refining method must be the same as that of the
+parent method it refines. This can be ensured by providing an explicit
+type signature equivalent to that of the parent (with renaming of
+parameters allowed) or by using an ellipsis (`...`) to indicate copying
+of the parent type signature. The following example illustrates that
+latter case.
+
+```
+module M0 {
+  method m(x: int) returns (r: int)
+}
+
+module M1 refines M0 {
+  method m...
+  {
+    return x + 1;
+  }
+}
+```
+
+As an example of semantic compatibility and increased determinism,
+consider the following example.
+
+```
+module M0 {
+  method m(x: int) returns (r: int)
+    ensures r > x;
+  {
+    var y :| y > x;
+    return y;
+  }
+}
+
+module M1 refines M0 {
+  method m(x: int) returns (r: int)
+  {
+    return x + 1;
+  }
+}
+```
+
+Here, `m` ensures that its result will be greater than `x`, but the
+version in `M0` returns an arbitrary value satisfying that
+postcondition. The version of `m` in `M1` is deterministic, and is
+required to satisfy all of the postconditions that its parent method
+declares.
+
+To introduce additional statements, the child method can include
+ellipses within the body to stand in for portions of code from the
+parent body. Dafny then attempts to merge the body of the child with the
+body of the parent by filling in the ellipses. For example:
+
+```
+module M0 {
+  method m(x: int) returns (r: int)
+    ensures r > x;
+  {
+    var y :| y > x;
+    return y;
+  }
+}
+
+module M1 refines M0 {
+  method m(x: int) returns (r: int)
+  {
+    ...;
+    return y + 1;
+  }
+}
+```
+
+Refinement of method bodies can go far beyond adding statements to the
+end, however. Full details of the algorithm used to perform the merge
+operation are available in [this
+paper](https://dl.acm.org/doi/10.1007/s00165-012-0254-3).
+
+A refining module can also introduce new `method` declarations or
+definitions that do not exist in the refinement parent.
 
 ## 21.6. Lemma declarations
 
-TODO
+As lemmas are (ghost) methods, the description of method refinement from
+the previous section also applies to lemma refinement.
```

Comment:
One question though: what happens to lemmas that are invalidated by the refinement? (Are they impossible?).

For example for your earlier example, a lemma (which may not be expressible in Dafny?) is "method `m(0)` can return `3`".  This lemma is wrong after refinement, since `m(0) == 1`.

Reply:
The rules for superimposition in a refining module are such that no proofs are invalidated. The refining module must prove that any statement replacements are valid, that added statements don't fail, and that any new `return` correctly establishes the enclosing method's postcondition.

Reply:
Oh, and new proof obligations (like newly introduced `ensures` clauses) are checked in the child.